File: year2021/day02/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  h, d = 0, 0

  for (c,v) in input:
    if c == 'forward':
      h += v
    elif c == 'up':
      d -= v
    elif c == 'down':
      d += v

  return h * d


def part_2():
  a, h, d = 0, 0, 0

  for (c,v) in input:
    if c == 'forward':
      h += v
      d += v * -a
    elif c == 'up':
      a += v
    elif c == 'down':
      a -= v

  return h * d


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [(dir, int(val)) for dir, val in (row.split() for row in file)]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2021/day02/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
  private static String[] DIRECTIONS;
  private static int[] RANGES;

  public static void main(String[] args) {
    try {
      setup(Files.readAllLines(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static Integer part1() {
    int height = 0;
    int depth = 0;

    for(int i = 0; i < RANGES.length; i++) {
      if (DIRECTIONS[i].equals("forward")) {
        height += RANGES[i];

      } else if (DIRECTIONS[i].equals("up")) {
        depth -= RANGES[i];

      } else if (DIRECTIONS[i].equals("down")) {
        depth += RANGES[i];

      }
    }

    return height * depth;
  }

  public static Integer part2() {
    int aim = 0;
    int height = 0;
    int depth = 0;

    for (int i = 0; i < RANGES.length; i++) {
      if (DIRECTIONS[i].equals("forward")) {
        height += RANGES[i];
        depth += RANGES[i] * (aim * -1);

      } else if (DIRECTIONS[i].equals("up")) {
        aim += RANGES[i];

      } else if (DIRECTIONS[i].equals("down")) {
        aim -= RANGES[i];

      }
    }

    return height * depth;
  }

  public static void setup(List<String> input) {
    DIRECTIONS = input.stream().map(line -> line.split(" ")[0]).toArray(String[]::new);
    RANGES = input.stream().map(line -> Integer.parseInt(line.split(" ")[1])).mapToInt(Integer::intValue).toArray();
  }
}